src/target/riscv/riscv-013.c

```diff
@@ -1063,34 +1065,47 @@ static int is_vector_reg(uint32_t gdb_regno)
 		gdb_regno == GDB_REGNO_VLENB;
 }
 
-static int prep_for_register_access(struct target *target, uint64_t *mstatus,
-		int regno)
+static int prep_for_register_access(struct target *target, riscv_reg_t *mstatus,
+		enum gdb_regno regno)
 {
-	if (is_fpu_reg(regno) || is_vector_reg(regno)) {
-		if (register_read_direct(target, mstatus, GDB_REGNO_MSTATUS) != ERROR_OK)
-			return ERROR_FAIL;
-		if (is_fpu_reg(regno) && (*mstatus & MSTATUS_FS) == 0) {
-			if (register_write_direct(target, GDB_REGNO_MSTATUS,
-						set_field(*mstatus, MSTATUS_FS, 1)) != ERROR_OK)
-				return ERROR_FAIL;
-		} else if (is_vector_reg(regno) && (*mstatus & MSTATUS_VS) == 0) {
-			if (register_write_direct(target, GDB_REGNO_MSTATUS,
-						set_field(*mstatus, MSTATUS_VS, 1)) != ERROR_OK)
-				return ERROR_FAIL;
-		}
-	} else {
-		*mstatus = 0;
+	if (!is_fpu_reg(regno) && !is_vector_reg(regno))
+		return ERROR_OK;
+
+	LOG_TARGET_DEBUG(target, "Preparing mstatus to access %s",
+			gdb_regno_name(regno));
+	if (target->state != TARGET_HALTED) {
+		LOG_TARGET_ERROR(target, "Unable to access %s: target not halted",
+				gdb_regno_name(regno));
+		return ERROR_FAIL;
 	}
```

Comment:
Addressed.

Reply:
@en-sc , @timsifive  I think that if access to a register requires some bits in mstatus register to be set we should always return an error if the target is not halted. The reason is that we don't have transactional semantics here. Meaning, that even if we've seen that FS is set it does not mean that it will still be set by the time  we actually  perform  the access. (and this can lead to nasty results as described here: https://github.com/riscv/riscv-openocd/pull/766)

Reply:
@aap-sc I agree, and I believe that is what the new code does. See https://github.com/riscv/riscv-openocd/pull/842/files#diff-701864e2be3210275085c84b8f6fa7f9eb6171a2a18ca71638be42a155c08c63R1101

Reply:
If I understand correctly, the issue is similar to the following:
1. target is running.
2. `mstatus` is read by OpenOCD, `mstatus.fs != 0`
3. the code being executed on the target sets `mstatus.fs = 0`
4. OpenOCD writes to a FPU register

This would create a situation similar to #766 (not exactly the same, since there the target was halted).
I see this as a valid justification for rejecting access to a register on a running target, that could potentially need a modification to `mstatus` , even when this modification is perceived unnecessary, based on the current value of `mstatus`.

However, some target could implement abstract command accesses to FPU registers on running target. There would be no way to leverage this feature by an OpenOCD user.

IMHO, the proper way to tackle this would be:
1. To fail in `prep_for_register_access` if target is not halted prior to reading `mstatus`
2. To execute an extra abstract access prior to prep_for_register_access on running target -- This would allow access to an FPU register on a running target if mstatus does not need to be changed, while also preserving the WA for  #766. I would like to address this in another commit since this is a major change in behavior that needs to be tested on hardware.

Reply:
Your solution of trying and then updating mstatus if necessary (assuming the target is currently running) is a good one. I also agree that it's large enough to warrant a separate PR.

However, from this PR please remove the is-running check. Otherwise this PR causes a regression where accessing FPRs while running no longer works when it used to work.

Reply:
Addressed.